### DISCUSSION TITLE
Creating macOS app with ADT on my Mini Mac

### DISCUSSION THREAD

**Discussion by danielkotzer** [airsdk/Adobe-Runtime-Support#2561] [General]
Hi,

I try to create a macOS app with ADT

I ended up creating a folder containing all the folders and files I need to create an app on mac, with a build.command file that is the batch file I need to double click on mac to run an ADT command line on mac.

the file contains this command:
/Applications/AIRSDK_MacOS/bin/adt -package -storetype KeychainStore -alias "3rd Party Mac Developer Application: My Company Name, Inc. (xxxxxxxxx)" -target bundle  -storetype KeychainStore -alias "3rd Party Mac Developer Application: My Company Name, Inc. (xxxxxxxxx)" AppName.app AppName-app.xml -extdir /Users/shai/Desktop/Mac/ane AppName.swf /Users/shai/Desktop/Mac/assets /Users/shai/Desktop/Mac/icon

I get this in return:
Last login: Thu Mar 30 22:41:08 on ttys002

/Users/shai/Desktop/Mac/build.command ; exit;                                  

shai@Shais-Mac-mini-2 ~ % /Users/shai/Desktop/Mac/build.command ; exit;

unexpected failure: null

java.lang.NullPointerException

         at com.adobe.air.ApplicationPackager.addExtensionDir(ApplicationPackager.java)

         at com.adobe.air.nai.NativePackager.addExtensionDir(NativePackager.java)

         at com.adobe.air.ADT.parseFileAndPathArguments(ADT.java)

         at com.adobe.air.ADT.parseDescriptorAndFileset(ADT.java)

         at com.adobe.air.ADT.parseInput(ADT.java)

         at com.adobe.air.ADT.parsePackage(ADT.java)

         at com.adobe.air.ADT.parseArgsAndGo(ADT.java)

         at com.adobe.air.ADT.run(ADT.java)

         at com.adobe.air.ADT.main(ADT.java)

 

Saving session...

...copying shared history...

...saving history...truncating history files...

...completed.

 

[Process completed]

What am I doing wrong?

**Comment by ajwfrost** [comment]
This looks like an error processing one of the ANEs within your '/Users/shai/Desktop/Mac/ane' folder. Depending on what version of macOS you're running on, I am wondering whether your build command may not have permissions to access that folder. Can you add a `ls /Users/shai/Desktop/Mac/ane` command in there and check that you see the expected results?

The function that it's complaining about is accessing a File object (the extension directory) and listing the files that are in them, and for each of those files it's checking for the filename ending in ".ane". So the only way we should get a null exception is if we're unable to read a filename (!) or if the input file is null, but I can't see any way in code that a null value could get in:
https://docs.oracle.com/javase/7/docs/api/java/io/File.html#getAbsoluteFile() - no hint that this can return null..

So if you can check the contents of that folder are accessible from the build command script and output what's in them, it may provide some hints. Other than that we would need to add a lot of custom debug in there for you as it's a very strange/unexpected error!

thanks

**Reply by danielkotzer** [reply]
did it got this result:
ls: /Users/shai/Desktop/Mac/ane: Operation not permitted
What do I do next?

**Reply by ajwfrost** [reply]
If you open up Settings -> Security and Privacy, do you see any message there about an attempt to access stuff? Sometimes if it blocks things, it will then add a message there to allow you to change the permissions.

Another thought might be if you have your command do some interactive prompt or similar, I'm wondering whether it's possible to run the "ls" command manually as I would normally expect an OS pop-up saying that it's trying to access things, and allowing you to say then and there that you accept it...

Only other thought might be to try putting everything in same folder - I think the script is already in the same place as the ane folder? but you could copy the adt script into that folder too (and edit it to set the path to the adt.jar file) - maybe that would help? (but I'm guessing now, Apple security changes fairly rapidly so I'm not 100% sure on this I'm afraid!)

thanks

**Comment by danielkotzer** [comment]
After playing with it for some time, I realized the problem, the reason it gave an error as if there is no access to the file, is because the path to the file was wrong, once I corrected the path to the file it worked.
I created a file name by the name: build.command
that contained this command: 
bash /Applications/AIRSDK_MacOS/bin/adt -package -storetype KeychainStore -alias "3rd Party Mac Developer Application: Company Name, Inc. (xxxxxxxx)" -target bundle  -storetype KeychainStore -alias "3rd Party Mac Developer Application: Company Name, Inc. (xxxxxxxx)" Desktop/Mac/AppName.app Desktop/Mac/AppName-app.xml -extdir Desktop/Mac/ane Desktop/Mac/AppName.swf Desktop/Mac/assets Desktop/Mac/icon

And it worked, the app file with the ANEs was created, didn't test the functionality of these ANEs yet, but the app is working the way it should be, more or less.

My question is why can't I use a relative path like I do when I create a batch file in windows? If for instance I put a file name in a command line in windows, in a batch file it will look for it on the location of the batch file, while with this command line I have to add the path for each file, like so: Desktop/Mac/AppName.swf Why? If I activate a batch file it has to assume a file without a path means it has to look for it in the same folder where the batch file is located!
Is there something I can do to make it so that no path means batch file path?

**Reply by ajwfrost** [reply]
Wouldn't that depend on how the batch file is executed? If you're able to put a filename as "Desktop/Mac/AppName.swf" and it's working, then this indicates that relative paths are working fine, they're just relative to your home folder...

So perhaps either execute the script as:
`cd ~/Desktop/Mac && sh ~/build.command`
or add in the start of it `pushd ~/Desktop/Mac`?

I think this is just a Windows vs Mac (or linux) thing in how they handle "working" folders...

thanks

**Comment by danielkotzer** [comment]
I have successfully created a macOS application on my mini mac with ADT using a *.command file I can simply double click to run, here is its content:
`#!/bin/bash
cd "$(dirname "$BASH_SOURCE")" || {
    echo "Error getting script directory" >&2
    exit 1
}
bash /Applications/AIRSDK_MacOS/bin/adt -package -storetype KeychainStore -alias "3rd Party Mac Developer Application: MyCompany, Inc. (xxxxxxxxxx)" -target bundle  -storetype KeychainStore -alias "3rd Party Mac Developer Application: MyCompany, Inc. (xxxxxxxxxx)" AppName.app AppName-app.xml -extdir ane AppName.swf assets icons`

then I sign it with another command and create a pkg file and upload it to mac app store successfully with Transporter, problem is I get one issue while uploading: the app is missing a provision profile which makes it impossible to use it for testflight.

I try to add this line to the command line but all I get is errors:
`-provisioning-profile AppName_Mac_Store.provisionprofile`

Where and how should I add it to the command line shown above?
